### PR TITLE
Moved select_pointer_type implementation into test-gen

### PR DESCRIPTION
**Pull request by thk123**
The base classs behaviour now leaves the type untouched. This requires no further testing so the original tests are now removed.

The `select_pointer_type` class is now passed in to allow configurable behaviour. However, on this branch the behaviour is trivial.

Corresponding test-gen PR: diffblue/test-gen#778

### CODE REVIEW

File: src/java_bytecode/java_object_factory.cpp [hunk 1]
```diff
@@ -139,6 +142,8 @@ class java_object_factoryt
     code_blockt &assignments,
     bool create_dynamic_objects,
     const pointer_typet &substitute_pointer_type);
+
+  std::shared_ptr<select_pointer_typet> pointer_type_selector;
```
Comment (cesaro) [inline src/java_bytecode/java_object_factory.cpp]: This new field is alone at the end of the class, while other fields are together at the beginning. Maybe this could be moved together with other fields (easier to read and find)? Would it be possible to document the field? What it is used for?

File: src/java_bytecode/java_object_factory.cpp [hunk 2]
```diff
@@ -993,6 +999,7 @@ void gen_nondet_init(
   bool create_dyn_objs,
   bool assume_non_null,
   size_t max_nondet_array_length,
+  std::shared_ptr<select_pointer_typet> pointer_type_selector,
```
Comment (cesaro) [inline src/java_bytecode/java_object_factory.cpp]: The new parameter of the function should be documented ;)

File: src/java_bytecode/java_object_factory.cpp [hunk 3]
```diff
@@ -913,7 +917,8 @@ exprt object_factory(
   bool allow_null,
   symbol_tablet &symbol_table,
   size_t max_nondet_array_length,
-  const source_locationt &loc)
+  const source_locationt &loc,
+  std::shared_ptr<select_pointer_typet> pointer_type_selector)
```
Comment (cesaro) [inline src/java_bytecode/java_object_factory.cpp]: This new function parameter should be documented above ;)

File: src/java_bytecode/java_object_factory.h [hunk 4]
```diff
@@ -10,18 +10,24 @@ Author: Daniel Kroening, kroening@kroening.com
...
 exprt object_factory(
   const typet &type,
   const irep_idt base_name,
   code_blockt &init_code,
   bool allow_null,
   symbol_tablet &symbol_table,
   size_t max_nondet_array_length,
-  const source_locationt &);
+  const source_locationt &,
+  std::shared_ptr<select_pointer_typet> pointer_type_selector);
```
Comment (cesaro) [inline src/java_bytecode/java_object_factory.h]: Should you have a default value for this new parameter, that provides the default implementation (the identity mapping) ? That would remove the additional parameter that you have added on all calls in this PR.

File: src/java_bytecode/java_entry_point.cpp [hunk 5]
```diff
@@ -137,7 +139,8 @@ exprt::operandst java_build_arguments(
   code_blockt &init_code,
   symbol_tablet &symbol_table,
   bool assume_init_pointers_not_null,
-  unsigned max_nondet_array_length)
+  unsigned max_nondet_array_length,
```
Comment (peterschrammel) [inline src/java_bytecode/java_entry_point.cpp]: There is a mixture of types here. `size_t` vs `unsigned`.

File: src/java_bytecode/java_object_factory.h [hunk 6]
```diff
@@ -10,18 +10,24 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_JAVA_BYTECODE_JAVA_OBJECT_FACTORY_H
 #define CPROVER_JAVA_BYTECODE_JAVA_OBJECT_FACTORY_H
 
+
```
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.h]: remove one blank line

File: src/java_bytecode/java_object_factory.h (hunk 4 above)
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.h]: Parameter default values should be avoided.

File: src/java_bytecode/java_object_factory.h [hunk 7]
```diff
@@ -10,18 +10,24 @@ Author: Daniel Kroening, kroening@kroening.com
...
 exprt object_factory(
   const typet &type,
   const irep_idt base_name,
   code_blockt &init_code,
   bool allow_null,
   symbol_tablet &symbol_table,
   size_t max_nondet_array_length,
-  const source_locationt &);
+  const source_locationt &,
```
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.h]: It looks inconsistent that this parameter is not named.

File: src/java_bytecode/select_pointer_type.cpp [hunk 8]
```diff
@@ -11,56 +11,15 @@
...
-#include <util/invariant.h>
-#include <util/namespace.h>
 #include <util/std_types.h>
 
-
-/// Select what type should be used for a given pointer type. If the class
-/// pointed to is abstract then we consider using a subclass of it.
+/// Select what type should be used for a given pointer type. For the base class
+/// we just use the supplied type. Dervied classes can override this behaviour
+/// to provide more sophisitcated type selection
```
Comment (peterschrammel) [inline src/java_bytecode/select_pointer_type.cpp]: sophisticated

File: src/java_bytecode/select_pointer_type.cpp [hunk 9]
```diff
@@ -11,56 +11,15 @@
...
-
-#include <util/invariant.h>
-#include <util/namespace.h>
 #include <util/std_types.h>
 
-
-/// Select what type should be used for a given pointer type. If the class
-/// pointed to is abstract then we consider using a subclass of it.
+/// Select what type should be used for a given pointer type. For the base class
+/// we just use the supplied type. Dervied classes can override this behaviour
```
Comment (peterschrammel) [inline src/java_bytecode/select_pointer_type.cpp]: Derived

File: src/java_bytecode/java_object_factory.h (hunk 4 above)
Comment (thk123) [inline src/java_bytecode/java_object_factory.h]: @peterschrammel @cesaro  How would you feel about an overloaded method that passes the identify one in?

Comment (cesaro) [inline src/java_bytecode/java_object_factory.h]: @peterschrammel what is the reason for that? (I'm 10m away from you if you want to replay orally)
@thk123 that would be OK for me

Comment (thk123) [inline src/java_bytecode/java_object_factory.h]: I have done this - I'm no longer sure about overloading the method - I feel like it has all the same draw backs of a default value (my main issue with it is can often get confusing resolution in C++ due to all of the 5 final parameters, though have distinct types, are all interchangeable).  I've done it as a standalone commit, but my feeling is I should drop it.

**Comment by thk123** [comment]
@peterschrammel @cesaro Changes applies as requested.

**Comment by thk123** [comment]
@peterschrammel @cesaro Let me know what you would like me to do regarding the overloaded versions of `object_factory` and `gen_nondet_init`

File: src/java_bytecode/java_object_factory.cpp [hunk 10]
```diff
@@ -57,6 +58,11 @@ class java_object_factoryt
   symbol_tablet &symbol_table;
   namespacet ns;
 
+  /// Resolves pointer types potentially using some heuristics for example
+  /// to replace pointers to interface types with pointers to concrete
+  /// implementations.
+  std::shared_ptr<select_pointer_typet> pointer_type_selector;
```
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.cpp]: Is there reason why this can't be a unique pointer?

File: src/java_bytecode/java_object_factory.cpp [hunk 11]
```diff
@@ -1026,3 +1044,49 @@ void gen_nondet_init(
 
   init_code.append(assignments);
 }
+
+/// Call object_factory with a default (identity) pointer_type_selector
```
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.cpp]: Does it make sense to actually provide an implementation for the identity selector? (and thus being able to use a reference member instead of a pointer)

Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Sorry I don't understand this comment - the `select_pointer_typet` is the identity selector. I will see if I can make it work with a reference or a `unique_ptr`

**Comment by thk123** [comment]
@peterschrammel Replaced with `const select_pointer_typet &` since doesn't need to be reassigned.

File: src/java_bytecode/java_object_factory.h (hunk 4 above)
Comment (cesaro) [inline src/java_bytecode/java_object_factory.h]: @thk123 I agree, maybe it was better not doing it. Apologies for my suggestion. I leave it up to you and the other reviewers whether it should get in or not.
